devito/operations/interpolators.py

```diff
@@ -173,42 +195,74 @@ def grid(self):
         return self.sfunction.grid
 
     @property
-    def _weights(self):
+    def r(self):
+        return self.sfunction.r
+
+    @memoized_meth
+    def _weights(self, subdomain=None):
         raise NotImplementedError
 
     @property
     def _gdims(self):
         return self.grid.dimensions
 
-    @property
-    def r(self):
-        return self.sfunction.r
-
     @cached_property
-    def _rdim(self):
+    def _cdim(self):
+        """Base CustomDimensions used to construct _rdim"""
         parent = self.sfunction._sparse_dim
         dims = [CustomDimension("r%s%s" % (self.sfunction.name, d.name),
                                 -self.r+1, self.r, 2*self.r, parent)
                 for d in self._gdims]
+        return dims
+
+    @memoized_meth
+    def _rdim(self, subdomain=None):
+        # If the interpolation operation is limited to a SubDomain,
+        # use the SubDimensions of that SubDomain
+        if subdomain:
+            gdims = tuple(subdomain.dimension_map[d] for d in self._gdims)
+        else:
+            gdims = self._gdims
 
         # Make radius dimension conditional to avoid OOB
         rdims = []
         pos = self.sfunction._position_map.values()
 
-        for (d, rd, p) in zip(self._gdims, dims, pos):
+        for (d, rd, p) in zip(gdims, self._cdim, pos):
             # Add conditional to avoid OOB
             lb = sympy.And(rd + p >= d.symbolic_min - self.r, evaluate=False)
             ub = sympy.And(rd + p <= d.symbolic_max + self.r, evaluate=False)
             cond = sympy.And(lb, ub, evaluate=False)
+
+            # Insert a check to catch cases where interpolation/injection is
+            # into an empty rank. This depends on the injection field or interpolated
+            # expression, and so must be inserted here.
+            if subdomain and subdomain.distributor.is_parallel:
+                rank_populated = subdomain.distributor.rank_populated
+                cond = sympy.And(rank_populated, cond)
+
             rdims.append(ConditionalDimension(rd.name, rd, condition=cond,
                                               indirect=True))
 
-        return DimensionTuple(*rdims, getters=self._gdims)
+        return DimensionTuple(*rdims, getters=gdims)
 
     def _augment_implicit_dims(self, implicit_dims, extras=None):
         if extras is not None:
+            # If variables are defined on a SubDomain of the Grid, then temprarily add
+            # their dimensions to self._gdims for the purpose of checking if extra
+            # dimensions have been included.
+            edims = []
+            for v in extras:
+                try:
+                    egrid = v.grid
+                    if egrid is not None and egrid.is_SubDomain:
```

Comment:
`egrid is not None` isn't needed

the line `egrid = v.grid` can probably be dropped too